### ISSUE TITLE
Use the ucf tool on Debian/Ubuntu to simplify upgrades

### ISSUE THREAD

**Issue by jlecour** [elastic/elasticsearch#26050]
**Describe the feature**:

When upgrading Elasticsearch with a .deb package, the least customization of the config file, brings a painful question from dpkg : should I use the new config file (with possibly good new stuff), keep using the current file or resolve this by hand ?

Many years ago (since wheezy) Debian has introduced **ucf**, a tool dedicated to help resolve this issue. The maintainer of the package has to do a little more work and has to document the upgraded config file better, but then ucf will help doing a three-way merge of the previous default file, the new config file and the customized one.

I'm no expert in this tol, but its usage is encouraged by the Debian team and it would change the upgrade process for the better with automatic resolution of most "conflicts".

**Elasticsearch version** (`bin/elasticsearch --version`): all

**Plugins installed**: N/A

**JVM version** (`java -version`): N/A

**OS version** (`uname -a` if on a Unix-like system): Debian/Ubuntu

**Comment by danielmitterdorfer** [comment]
Thanks for the suggestion @jlecour! @rjernst, @nik9000, @jasontedor do you have any thoughts on this?

**Comment by jasontedor** [comment]
I have a different understanding of the purpose of `ucf` from what seems to be the case here. My understanding is that `ucf` is primarily for use cases where configuration files are generated during post-install; that is, they are *not* part of the shipped package, so can not be marked as configuration files, so do not receive the automatic conflict resolution that `dpkg` provides. Instead, `ucf` enables this functionality for invocation in the post-install scripts for such configuration files. We do not need this, we do not generate configuration files in post-install instead shipping all configuration files as part of the shipped package. I do not see a need for us to transition away from the automatic management that `dpkg` provides; it's simple, and well-understood.

Here are the changes in shipped configuration file during the 5.x series:
 - from 5.0 to 5.5, the elasticsearch.yml only saw comment changes; these are easily resolved
 - from 5.0 to 5.5, the jvm.options file removed `-XX:+DisableExplicitGC`, added `-Xss1m`, added `-Dio.netty.recycler.maxCapacityPerThread=0`, added commented-out GC log rotation, and a few comments were modified
 - from 5.0 to 5.5, the log4j2.properties changed to support a system properties that we set for controlling the log files

I don't think we'll see many changes to the elasticsearch.yml, so the chance of conflicts there is minimal and they are going to be easy to resolve.

Other than heap size, the vast majority of users should accept the JVM defaults that we ship with, so there should rarely be conflicts and they should all be straightforward to resolve.

For the log4j2.properties, I can see some users customizing these but I think these conflicts are rare (we do not make changes to this file frequently), and would be easy to resolve.

While `ucf` can be used for configuration files like those that we ship (i.e., not just ones produced during post-installation), the only advantage that it buys us a three-way merge for conflicts. Given the above, I do not see the need to add `ucf` for this. Additionally, I'm keen to keep the Debian distributions as close as possible.

**Comment by jlecour** [comment]
Hi @danielmitterdorfer, thank you for you answer.

Providing a way to resolve 3-way merges of configuration files (like the main YAML file) is exactly why I brought this topic to attention.

Of course, confilcts can be resolved manually when dpkg finds that the file has changed from the default (which always happens, at least to customize the network and the cluster/node names).
Each time the config file has a change of comments, the user has to parse the whole file visually of programmaticaly to find if something meaningful has changed or not.
With ucf, there is still an interactive choice to make ("resolve 3-way merge") but the resolution itself is automatic in a typical situation. It might even become an option for non-interactive frontends.

I've tried to do this and the results seems really good to me.
1. i've download the 5.0 package and extracted the YAML file
2. i've done the same for the 5.5 package
3. I've copied the 5.0 config and customized it, for a few common variables.

With the help of ucf, I have a 5.5 config with all my customizations, but programmed automatically
I didn't need to convert the default files to any other format. Generating a config file at post-install time doesn't mean that they have to contains variables or any logic.

Fetch the packages and initialize ucf for the config file
~~~
# cd /tmp
# wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-5.0.0.deb
# mkdir es50
# cd es50
# ar x ../elasticsearch-5.0.0.deb
# tar xzf data.tar.gz
# cd /tmp
# wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-5.5.1.deb
# mkdir es55
# cd es55
# ar x ../elasticsearch-5.5.1.deb
# tar xzf data.tar.gz
#  ucf -v --three-way /tmp/es50/etc/elasticsearch/elasticsearch.yml /tmp/elasticsearch.yml
: The new file is /tmp/es50/etc/elasticsearch/elasticsearch.yml
: The Destination file is /tmp/elasticsearch.yml
: The Source directory is /tmp/es50/etc/elasticsearch
: The State directory is /var/lib/ucf
Creating config file /tmp/elasticsearch.yml with new version
~~~

customize the config file
~~~.diff
--- es50/etc/elasticsearch/elasticsearch.yml	2016-10-26 00:40:43.000000000 -0400
+++ elasticsearch.yml	2017-08-10 16:52:16.835867854 -0400
@@ -17,0 +18 @@
+cluster.name: my-cluster
@@ -23,0 +25 @@
+node.name: my-node
@@ -55,0 +58 @@
+network.host: [_local_];
~~~

Let's update the config file :
~~~
# ucf -v --three-way /tmp/es55/etc/elasticsearch/elasticsearch.yml /tmp/elasticsearch.yml
: The new file is /tmp/es55/etc/elasticsearch/elasticsearch.yml
: The Destination file is /tmp/elasticsearch.yml
: The Source directory is /tmp/es55/etc/elasticsearch
: The State directory is /var/lib/ucf
The hash file exists
egrep [[:space:]]\/tmp\/elasticsearch\.yml$ /var/lib/ucf/hashfile
ee49e0f90de02a818891a18ebf91c35a  /tmp/elasticsearch.yml
Merging changes into the new version
(egrep -v "[[:space:]]\/tmp\/elasticsearch\.yml$" "/var/lib/ucf/hashfile";
31655f8d9bce94581724510ebc90ed75  /tmp/elasticsearch.yml
~~~

And now the file is properly updated/merged as you can in the diff between the 5.0 config and the one after the upgrade :
~~~.diff
--- /tmp/es50/etc/elasticsearch/elasticsearch.yml	2016-10-26 00:40:43.000000000 -0400
+++ /tmp/elasticsearch.yml	2017-08-10 16:58:19.771858018 -0400
@@ -10,2 +10,2 @@
-# Please see the documentation for further information on configuration options:
-# <http://www.elastic.co/guide/en/elasticsearch/reference/current/setup-configuration.html>
+# Please consult the documentation for further information on configuration options:
+# https://www.elastic.co/guide/en/elasticsearch/reference/index.html
@@ -17,0 +18 @@
+cluster.name: my-cluster
@@ -23,0 +25 @@
+node.name: my-node
@@ -55,0 +58 @@
+network.host: [_local_];
@@ -61,2 +64 @@
-# For more information, see the documentation at:
-# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html>
+# For more information, consult the network module documentation.
@@ -71 +73 @@
-# Prevent the "split brain" by configuring the majority of nodes (total number of nodes / 2 + 1):
+# Prevent the "split brain" by configuring the majority of nodes (total number of master-eligible nodes / 2 + 1):
@@ -75,2 +77 @@
-# For more information, see the documentation at:
-# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery.html>
+# For more information, consult the zen discovery module documentation.
@@ -84,2 +85 @@
-# For more information, see the documentation at:
-# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-gateway.html>
+# For more information, consult the gateway module documentation.
@@ -88,4 +87,0 @@
-#
-# Disable starting multiple nodes on a single system:
-#
-#node.max_local_storage_nodes: 1
~~~

So, I think ucf does the job pretty well, with significantly less attention/action from the user's perspective. I can't speak from first principle about the amount of work on the packaging side, but it seems reasonable.

**Comment by danielmitterdorfer** [comment]
While I agree that `ucf` seems nice and helpful, @jasontedor has reasoned that the configuration changes only seldom and in addition we want the Debian and RPM distributions to stay as close as possible for maintenance reasons. Hence I'm closing it for the time being and in case we'll pick it up in the future, we can reopen.
